persistence/rocksdb/src/main/java/org/infinispan/persistence/rocksdb/RocksDBStore.java

```diff
@@ -187,271 +190,268 @@ protected RocksDB openDatabase(Path location, Options options) throws RocksDBExc
-        } catch (InterruptedException e) {
-            throw new PersistenceException("Cannot acquire semaphore", e);
-        }
-        try {
+    public CompletionStage<Void> stop() {
+        return blockingManager.runBlocking(() -> {
             handler.close();
             expiredDb.close();
-        } finally {
-            stopped = true;
-            semaphore.release(Integer.MAX_VALUE);
-        }
-    }
-
-    @Override
-    public void destroy() {
-        stop();
-        Util.recursiveFileRemove(getLocation().toFile());
-        Util.recursiveFileRemove(getExpirationLocation().toFile());
+        }, "rocksdb-stop");
     }
 
     @Override
-    public boolean isAvailable() {
-        return getLocation().toFile().exists() && getExpirationLocation().toFile().exists();
+    public Set<Characteristic> characteristics() {
+        return EnumSet.of(Characteristic.BULK_READ, Characteristic.EXPIRATION, Characteristic.SEGMENTABLE);
     }
 
     @Override
-    public void clear() {
-        handler.clear(null);
+    public CompletionStage<Boolean> isAvailable() {
+        return blockingManager.supplyBlocking(() -> getLocation().toFile().exists() && getExpirationLocation().toFile().exists(),
+                "rocksdb-available");
     }
 
     @Override
-    public void clear(IntSet segments) {
-        handler.clear(segments);
+    public CompletionStage<Void> clear() {
+        return handler.clear();
     }
 
     @Override
-    public int size() {
-        return handler.size(null);
-    }
-
-    @Override
-    public int size(IntSet segments) {
+    public CompletionStage<Long> size(IntSet segments) {
         return handler.size(segments);
     }
 
     @Override
-    public boolean contains(Object key) {
-        return handler.contains(-1, key);
+    public CompletionStage<Long> approximateSize(IntSet segments) {
+        return handler.approximateSize(segments);
     }
 
     @Override
-    public boolean contains(int segment, Object key) {
-        return handler.contains(segment, key);
-    }
-
-    @Override
-    public Publisher<K> publishKeys(Predicate<? super K> filter) {
-        return handler.publishKeys(null, filter);
+    public CompletionStage<Boolean> containsKey(int segment, Object key) {
+        // This might be able to use RocksDB#keyMayExist - but API is a bit flaky
+        return load(segment, key)
+                .thenApply(Objects::nonNull);
     }
 
     @Override
     public Publisher<K> publishKeys(IntSet segments, Predicate<? super K> filter) {
-        return handler.publishKeys(segments, filter);
-    }
-
-    @Override
-    public Publisher<MarshallableEntry<K, V>> entryPublisher(Predicate<? super K> filter, boolean fetchValue, boolean fetchMetadata) {
-        return handler.publishEntries(null, filter, fetchValue, fetchMetadata);
+        return Flowable.fromPublisher(handler.publishEntries(segments, filter, false))
+                .map(MarshallableEntry::getKey);
```

Comment:
As `publishEntries` utilises the `RocksEntryIterator` we're also unnecessarily unmarshalling the value. In this case I think we can simply add a boolean to the `RocksEntryIterator` constructor and pass the `fetchValue` parameter to it.

Reply:
I looked closely at this one, unfortunately due to the contract of keyPublisher with expiration and the way values are stored we have to always unmarshall the value to get the metadata, so I just consolidated it.